Correction of issue platisd#6 - Error in print(float) function

MarcBousquet opened the pull request:
numbers like 1.0123 or 1.00123 were printed 1.123, having the 0s next to the coma all skipped.

From the correction proposed in the issue discussion, i changed the `&& ret` parts with `|| out_of_bounds` to respect the propagation of the out_of_bounds state.

MarcBousquet committed:
Update Nokia_LCD.cpp

Correction of issue #6 - Error in print(float) function
numbers like 1.0123 or 1.00123 were printed 1.123, having the 0s next to the coma all skipped.

MarcBousquet committed:
Merge pull request #1 from MarcBousquet/MarcBousquet-issue#6

Update Nokia_LCD.cpp

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -263,11 +263,27 @@ bool Nokia_LCD::print(unsigned long number) {
     return print(number_as_string);
 }
 
+//----
+// 27-Nov-2020 - Correction of issue #6 - Error in print(float) function
+//
+// In previous release, numbers like 1.0123 or 1.00123 were printed 1.123,
+// having the 0s next to the coma skipped.
+//----
```
Please remove :)
We have `git` for keeping the history and the change log.

MarcBousquet replied:
Done

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -263,11 +263,27 @@ bool Nokia_LCD::print(unsigned long number) {
+	out_of_bounds = print(static_cast<long>(number)) || out_of_bounds;
+	out_of_bounds = print(".") || out_of_bounds;
+	
+	for(int d=decimals-1 ; 0<=d; d-- ) {
+		int num = fractional/pow(10, d);
```
Can you please fix the spacing? it should look something like:

```cpp
	for(int d = decimals - 1; 0 <= d; d--) {
		int num = fractional / pow(10, d);
```

MarcBousquet replied:
Done

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -263,11 +263,27 @@ bool Nokia_LCD::print(unsigned long number) {
+	
+	for(int d=decimals-1 ; 0<=d; d-- ) {
+		int num = fractional/pow(10, d);
+		if (num) break; //print only leading 0s
+		out_of_bounds = print(num) || out_of_bounds;
```
Can you be more explicit with the comparison of numeric values?
E.g.
```cpp
// print only leading 0s
if (num == 0) {
    out_of_bounds = print(num) || out_of_bounds;
}
```

MarcBousquet replied:
I changed the if statement to a more readable form.

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -263,11 +263,27 @@ bool Nokia_LCD::print(unsigned long number) {
     long fractional = pow(10.0, decimals) * modf(number, &integral);
+    bool out_of_bounds = false;
 
-    return print(static_cast<long>(number)) || print(".") || print(fractional);
```
The introduced code isn't equivalent to the previous one in regards to what happens when we get an out of bounds error.

Specifically, before, if one of those `print` statements would return `true`, then the entire expression would be lazily evaluated to `true` and the remaining statements would never execute. In the suggested case, the out of bounds flag will be correctly raised, however, the out of bounds text will keep being printed.
I don't remember what actually happens in that case and whether the cursor goes back to the beginning but this can potentially be a problem.

MarcBousquet replied:
You're right, corrected.

platisd replied:
Nice fix! Took me a second to realize how `out_of_bounds` is now used but it makes sense! :grin:

platisd reviewed (changes requested):
Looks good! There are some formatting remarks but mainly there's a logic issue in regards to what happens when there's an intermittent out of bounds error. ideally we shouldn't continue printing but instead return `true` directly.

More details in my comments in the code.

MarcBousquet committed:
Out of bounds behavior and cleanings

out_of_bounds Variable is placed before || so that no more digit is written if previous print raised out of bounds.
Some code cleanings and comments

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -263,11 +263,27 @@ bool Nokia_LCD::print(unsigned long number) {
+    out_of_bounds = out_of_bounds || print(".");
+
+    //prints any leading 0s after the dot
+    for(int d = decimals - 1 ; out_of_bounds || (0 <= d); d-- ) {
+        int num = fractional/pow(10, d);
+        if (num == 0) 
+ 	   out_of_bounds = out_of_bounds || print(num);
+	else
+           break; //no more 0s to print so break
+    }
```
This is technically correct and saves us a couple of loops if things have gone out of bounds.
However, wasting a few loops in the name of simplicity is I believe preferred in this occasion. The code would now look like this:

```cpp
    for (int d = decimals - 1; d >= 0; d--) {
        int num = fractional / pow(10, d);
        if (num == 0) {
            out_of_bounds = out_of_bounds || print(num);
        } else {
            break;
       }
    }
```

I also added some braces after the `if-else` statement and formatted the spaces to be consistent with the rest of the code.

MarcBousquet replied:
Ok. Done.

platisd reviewed (changes requested):
Almost there! :smile:

MarcBousquet committed:
Remove optimization and code Channing

Remove loop count optimization for better choice reading.
Few additional baskets.

platisd reviewed (approved):
Looks good, thanks!
